// node-typedstream/dist/types/foundation.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var NSObject_1, NSData_1, NSMutableData_1, NSDate_1, NSString_1, NSMutableString_1, NSAttributedString_1, NSMutableAttributedString_1, NSValue_1, NSNumber_1, NSArray_1, NSMutableArray_1, NSSet_1, NSMutableSet_1, NSDictionary_1, NSMutableDictionary_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.NSMutableDictionary = exports.NSDictionary = exports.NSMutableSet = exports.NSSet = exports.NSMutableArray = exports.NSArray = exports.NSNumber = exports.NSValue = exports.NSMutableAttributedString = exports.NSAttributedString = exports.NSMutableString = exports.NSString = exports.NSDate = exports.NSMutableData = exports.NSData = exports.NSObject = void 0;
const archiver_1 = require("../archiver");
const known_types_1 = require("./known_types");
let NSObject = NSObject_1 = class NSObject extends known_types_1.KnownArchivedObject {
    static initFromUnarchiver(unarchiver, archivedClass) {
        if (archivedClass.version != 0) {
            throw new EvalError(`Unsupported version: ${archivedClass.version}`);
        }
        return new NSObject_1();
    }
    ;
};
NSObject = NSObject_1 = __decorate([
    (0, known_types_1.archivedClass)("NSObject")
], NSObject);
exports.NSObject = NSObject;
let NSData = NSData_1 = class NSData extends NSObject {
    constructor(data) {
        super();
        this.data = data;
    }
    static initFromUnarchiver(unarchiver, archivedClass) {
        if (archivedClass.version != 0) {
            throw new EvalError(`Unsupported version: ${archivedClass.version}`);
        }
        return new NSData_1(Buffer.from(unarchiver.decodeDataObject()));
    }
    ;
};
NSData = NSData_1 = __decorate([
    (0, known_types_1.archivedClass)("NSData")
], NSData);
exports.NSData = NSData;
let NSMutableData = NSMutableData_1 = class NSMutableData extends NSData {
    static initFromUnarchiver(unarchiver, archivedClass) {
        if (archivedClass.version != 0) {
            throw new EvalError(`Unsupported version: ${archivedClass.version}`);
        }
        return new NSMutableData_1(unarchiver.decodeDataObject());
    }
    ;
};
NSMutableData = NSMutableData_1 = __decorate([
    (0, known_types_1.archivedClass)("NSMutableData")
], NSMutableData);
exports.NSMutableData = NSMutableData;
let NSDate = NSDate_1 = class NSDate extends NSObject {
    constructor(absoluteReferenceDateOffset) {
        super();
        this.absoluteReferenceDateOffset = absoluteReferenceDateOffset;
    }
    get value() {
        return new Date(NSDate_1.ABSOLUTE_REFERENCE_DATA + (1000 * this.absoluteReferenceDateOffset));
    }
    static initFromUnarchiver(unarchiver, archivedClass) {
        if (archivedClass.version != 0) {
            throw new EvalError(`Unsupported version: ${archivedClass.version}`);
        }
        return new NSDate_1(unarchiver.decodeValueOfType('d'));
    }
    ;
};
NSDate.ABSOLUTE_REFERENCE_DATA = Date.UTC(2001, 1, 1);
NSDate = NSDate_1 = __decorate([
    (0, known_types_1.archivedClass)("NSDate")
], NSDate);
exports.NSDate = NSDate;
let NSString = NSString_1 = class NSString extends NSObject {
    constructor(value) {
        super();
        this.string = value;
    }
    static initFromUnarchiver(unarchiver, archivedClass) {
        if (archivedClass.version != 1) {
            throw new EvalError(`Unsupported version: ${archivedClass.version}`);
        }
        return new NSString_1(unarchiver.decodeValueOfType('+').toString('utf8'));
    }
    ;
};
NSString = NSString_1 = __decorate([
    (0, known_types_1.archivedClass)("NSString")
], NSString);
exports.NSString = NSString;
let NSMutableString = NSMutableString_1 = class NSMutableString extends NSString {
    static initFromUnarchiver(unarchiver, archivedClass) {
        if (archivedClass.version != 1) {
            throw new EvalError(`Unsupported version: ${archivedClass.version}`);
        }
        return new NSMutableString_1(unarchiver.decodeValueOfType('+').toString('utf8'));
    }
    ;
};
NSMutableString = NSMutableString_1 = __decorate([
    (0, known_types_1.archivedClass)("NSMutableString")
], NSMutableString);
exports.NSMutableString = NSMutableString;
let NSAttributedString = NSAttributedString_1 = class NSAttributedString extends NSString {
    constructor(value, runs) {
        super();
        this.string = value;
        this.runs = runs;
    }
    static readString(unarchiver) {
        return unarchiver.decodeTypedValues().values[0].string;
    }
    static readRange(unarchiver) {
        const range = unarchiver.decodeTypedValues();
        return {
            reference: range.values[0],
            length: range.values[1],
        };
    }
    static readAttributeValue(unarchiver) {
        const dict = unarchiver.decodeTypedValues().values[0].contents;
        const attributeValue = {};
        for (const [key, value] of dict.entries()) {
            if (value instanceof NSData) {
                if (unarchiver.binaryDecoding != archiver_1.Unarchiver.BinaryDecoding.none) {
                    const data = value.data;
                    try {
                        const decoded = archiver_1.Unarchiver.open(data).decodeAll();
                        if (decoded.length == 1) {
                            attributeValue[key.string] = decoded[0];
                        }
                        else {
                            attributeValue[key.string] = decoded;
                        }
                    }
                    catch (e) {
                        if (unarchiver.binaryDecoding == archiver_1.Unarchiver.BinaryDecoding.all) {
                            attributeValue[key.string] = data;
                        }
                    }
                }
            }
            else if (value instanceof NSString) {
                attributeValue[key.string] = value.string;
            }
            else {
                attributeValue[key.string] = value.value;
            }
        }
        return attributeValue;
    }
    static readAttributes(unarchiver, length) {
        const runs = [];
        let index = 0;
        let sharedAttributeValues = {};
        while (index < length) {
            const range = this.readRange(unarchiver);
            if (!(range.reference in sharedAttributeValues)) {
                sharedAttributeValues[range.reference] = this.readAttributeValue(unarchiver);
            }
            const attributeValue = sharedAttributeValues[range.reference];
            runs.push({
                range: [index, range.length],
                attributes: attributeValue,
            });
            index += range.length;
        }
        return runs;
    }
    static initFromUnarchiver(unarchiver, archivedClass) {
        if (archivedClass.version != 0) {
            throw new EvalError(`Unsupported version: ${archivedClass.version}`);
        }
        const value = this.readString(unarchiver);
        return new NSAttributedString_1(value, this.readAttributes(unarchiver, value.length));
    }
    ;
};
NSAttributedString = NSAttributedString_1 = __decorate([
    (0, known_types_1.archivedClass)("NSAttributedString")
], NSAttributedString);
exports.NSAttributedString = NSAttributedString;
let NSMutableAttributedString = NSMutableAttributedString_1 = class NSMutableAttributedString extends NSAttributedString {
    static initFromUnarchiver(unarchiver, archivedClass) {
        if (archivedClass.version != 0) {
            throw new EvalError(`Unsupported version: ${archivedClass.version}`);
        }
        const value = this.readString(unarchiver);
        return new NSMutableAttributedString_1(value, this.readAttributes(unarchiver, value.length));
    }
    ;
};
NSMutableAttributedString = NSMutableAttributedString_1 = __decorate([
    (0, known_types_1.archivedClass)("NSMutableAttributedString")
], NSMutableAttributedString);
exports.NSMutableAttributedString = NSMutableAttributedString;
let NSValue = NSValue_1 = class NSValue extends NSObject {
    constructor(typeEncoding, value) {
        super();
        this.typeEncoding = typeEncoding;
        this.value = value;
    }
    static initFromUnarchiver(unarchiver, archivedClass) {
        if (archivedClass.version != 0) {
            throw new EvalError(`Unsupported version: ${archivedClass.version}`);
        }
        const typeEncoding = unarchiver.decodeValueOfType('*');
        const value = unarchiver.decodeValueOfType(typeEncoding);
        return new NSValue_1(typeEncoding, value);
    }
    ;
};
NSValue = NSValue_1 = __decorate([
    (0, known_types_1.archivedClass)("NSValue")
], NSValue);
exports.NSValue = NSValue;
let NSNumber = NSNumber_1 = class NSNumber extends NSValue {
    static initFromUnarchiver(unarchiver, archivedClass) {
        if (archivedClass.version != 0) {
            throw new EvalError(`Unsupported version: ${archivedClass.version}`);
        }
        const typeEncoding = unarchiver.decodeValueOfType('*');
        const value = unarchiver.decodeValueOfType(typeEncoding);
        return new NSNumber_1(typeEncoding, value);
    }
    ;
};
NSNumber = NSNumber_1 = __decorate([
    (0, known_types_1.archivedClass)("NSNumber")
], NSNumber);
exports.NSNumber = NSNumber;
let NSArray = NSArray_1 = class NSArray extends NSObject {
    constructor(elements) {
        super();
        this.elements = elements;
    }
    static initFromUnarchiver(unarchiver, archivedClass) {
        if (archivedClass.version != 0) {
            throw new EvalError(`Unsupported version: ${archivedClass.version}`);
        }
        const count = unarchiver.decodeValueOfType('i');
        if (count < 0) {
            throw new EvalError(`NSArray element count cannot be negative: ${count}`);
        }
        const elements = [];
        for (let i = 0; i < count; i++) {
            elements.push(unarchiver.decodeValueOfType('@'));
        }
        return new NSArray_1(elements);
    }
    ;
};
NSArray = NSArray_1 = __decorate([
    (0, known_types_1.archivedClass)("NSArray")
], NSArray);
exports.NSArray = NSArray;
let NSMutableArray = NSMutableArray_1 = class NSMutableArray extends NSArray {
    static initFromUnarchiver(unarchiver, archivedClass) {
        if (archivedClass.version != 0) {
            throw new EvalError(`Unsupported version: ${archivedClass.version}`);
        }
        const count = unarchiver.decodeValueOfType('i');
        if (count < 0) {
            throw new EvalError(`NSMutableArray element count cannot be negative: ${count}`);
        }
        const elements = [];
        for (let i = 0; i < count; i++) {
            elements.push(unarchiver.decodeValueOfType('@'));
        }
        return new NSMutableArray_1(elements);
    }
    ;
};
NSMutableArray = NSMutableArray_1 = __decorate([
    (0, known_types_1.archivedClass)("NSMutableArray")
], NSMutableArray);
exports.NSMutableArray = NSMutableArray;
let NSSet = NSSet_1 = class NSSet extends NSObject {
    constructor(elements) {
        super();
        this.elements = elements;
    }
    static initFromUnarchiver(unarchiver, archivedClass) {
        if (archivedClass.version != 0) {
            throw new EvalError(`Unsupported version: ${archivedClass.version}`);
        }
        const count = unarchiver.decodeValueOfType('I');
        const elements = [];
        for (let i = 0; i < count; i++) {
            elements.push(unarchiver.decodeValueOfType('@'));
        }
        return new NSSet_1(elements);
    }
    ;
};
NSSet = NSSet_1 = __decorate([
    (0, known_types_1.archivedClass)("NSSet")
], NSSet);
exports.NSSet = NSSet;
let NSMutableSet = NSMutableSet_1 = class NSMutableSet extends NSSet {
    static initFromUnarchiver(unarchiver, archivedClass) {
        if (archivedClass.version != 0) {
            throw new EvalError(`Unsupported version: ${archivedClass.version}`);
        }
        const count = unarchiver.decodeValueOfType('I');
        const elements = [];
        for (let i = 0; i < count; i++) {
            elements.push(unarchiver.decodeValueOfType('@'));
        }
        return new NSMutableSet_1(elements);
    }
    ;
};
NSMutableSet = NSMutableSet_1 = __decorate([
    (0, known_types_1.archivedClass)("NSMutableSet")
], NSMutableSet);
exports.NSMutableSet = NSMutableSet;
let NSDictionary = NSDictionary_1 = class NSDictionary extends NSObject {
    constructor(contents) {
        super();
        this.contents = contents;
    }
    static initFromUnarchiver(unarchiver, archivedClass) {
        if (archivedClass.version != 0) {
            throw new EvalError(`Unsupported version: ${archivedClass.version}`);
        }
        const count = unarchiver.decodeValueOfType('i');
        if (count < 0) {
            throw new EvalError(`NSDictionary element count cannot be negative: ${count}`);
        }
        const contents = new Map();
        for (let i = 0; i < count; i++) {
            const key = unarchiver.decodeValueOfType('@');
            const value = unarchiver.decodeValueOfType('@');
            contents.set(key, value);
        }
        return new NSDictionary_1(contents);
    }
    ;
};
NSDictionary = NSDictionary_1 = __decorate([
    (0, known_types_1.archivedClass)("NSDictionary")
], NSDictionary);
exports.NSDictionary = NSDictionary;
let NSMutableDictionary = NSMutableDictionary_1 = class NSMutableDictionary extends NSDictionary {
    static initFromUnarchiver(unarchiver, archivedClass) {
        if (archivedClass.version != 0) {
            throw new EvalError(`Unsupported version: ${archivedClass.version}`);
        }
        const count = unarchiver.decodeValueOfType('i');
        if (count < 0) {
            throw new EvalError(`NSMutableDictionary element count cannot be negative: ${count}`);
        }
        const contents = new Map();
        for (let i = 0; i < count; i++) {
            const key = unarchiver.decodeValueOfType('@');
            const value = unarchiver.decodeValueOfType('@');
            contents.set(key, value);
        }
        return new NSMutableDictionary_1(contents);
    }
    ;
};
NSMutableDictionary = NSMutableDictionary_1 = __decorate([
    (0, known_types_1.archivedClass)('NSMutableDictionary')
], NSMutableDictionary);
exports.NSMutableDictionary = NSMutableDictionary;
