// node-typedstream/dist/types/known_types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.structClass = exports.structClassesByEncoding = exports.KnownStruct = exports.archivedClass = exports.archivedClassesByName = exports.KnownArchivedObject = void 0;
const encodings_1 = require("../encodings");
class KnownArchivedObject {
    static initFromUnarchiver(unarchiver, archivedClass) {
        throw new Error("Method not implemented");
    }
    ;
}
exports.KnownArchivedObject = KnownArchivedObject;
exports.archivedClassesByName = new Map();
function archivedClass(archivedName, consumeEnd = false) {
    return (target) => {
        target.archivedName = archivedName;
        target.consumeEnd = consumeEnd;
        exports.archivedClassesByName.set(archivedName, target);
    };
}
exports.archivedClass = archivedClass;
class KnownStruct {
    constructor(fields) { }
}
exports.KnownStruct = KnownStruct;
exports.structClassesByEncoding = new Map();
function structClass(structName, fieldEncodings) {
    return (target) => {
        target.structName = structName;
        target.fieldEncodings = fieldEncodings;
        target.encoding = (0, encodings_1.buildStructEncoding)(fieldEncodings, structName);
        exports.structClassesByEncoding.set(target.encoding, target);
    };
}
exports.structClass = structClass;
